state/protocol/protocol_state/state/mutable_protocol_state_test.go

```diff
@@ -0,0 +1,650 @@
+		_, dbUpdates, err := s.mutableState.EvolveState(s.candidate.ParentID, s.candidate.View, []*flow.Seal{})
+		require.Error(s.T(), err)
+		require.True(s.T(), dbUpdates.IsEmpty())
+	})
+
+	failingFactory.On("Create", s.candidate.View, s.candidate.ParentID, &s.parentState, &s.evolvingState).Return(nil, exception).Once()
+	s.Run("failing factory is first", func() {
+		s.kvStateMachineFactories[0], s.kvStateMachineFactories[1] = failingFactory, workingFactory //nolint:govet
+		_, dbUpdates, err := s.mutableState.EvolveState(s.candidate.ParentID, s.candidate.View, []*flow.Seal{})
+		require.Error(s.T(), err)
+		require.True(s.T(), dbUpdates.IsEmpty())
+	})
+}
+
+// Test_StateMachineProcessingServiceEventsFails verifies that errors are escalated to the caller that any of the sub-state machines return
+// when processing the service events. Specifically, the state machine should handle invalid events internally and only escalate internal
+// errors in case of an irrecoverable problem. The MutableProtocolState should _not_ interpret errors from the sub-state machines as
+// signs of invalid service events.
+// Protocol Convention:
+//   - The Orthogonal Store State Machines have a 3-step process to evolve their respective sub-states:
+//     (i) construction (via the injected factories)
+//     (ii) processing ordered service events from sealed results (in `EvolveState` call)
+//     (iii) in the `Build` step, the state machine assembles their resulting sub-state and the corresponding database operations
+//     to index and persist its substate.
+//   - The protocol convention is that `MutableProtocolState` executes first step (i) on all state machines, then step (ii) and lastly (iii)
+//
+// This test also verifies that the `MutableProtocolState` does not engage in step (iii) before the completing step (ii) on all state machines.
+func (s *StateMutatorSuite) Test_StateMachineProcessingServiceEventsFails() {
+	workingStateMachine := *protocol_statemock.NewOrthogonalStoreStateMachine[protocol.KVStoreReader](s.T())
+	workingStateMachine.On("EvolveState", mock.MatchedBy(emptySlice[flow.ServiceEvent]())).Return(nil).Once()
+
+	exception := errors.New("exception")
+	failingStateMachine := *protocol_statemock.NewOrthogonalStoreStateMachine[protocol.KVStoreReader](s.T())
+	failingStateMachine.On("EvolveState", mock.MatchedBy(emptySlice[flow.ServiceEvent]())).Return(exception).Once()
+
+	s.Run("failing state machine is last", func() {
+		s.kvStateMachines[0], s.kvStateMachines[1] = workingStateMachine, failingStateMachine //nolint:govet
+		_, dbUpdates, err := s.mutableState.EvolveState(s.candidate.ParentID, s.candidate.View, []*flow.Seal{})
+		require.Error(s.T(), err)
+		require.False(s.T(), protocol.IsInvalidServiceEventError(err))
+		require.True(s.T(), dbUpdates.IsEmpty())
+	})
+
+	failingStateMachine.On("EvolveState", mock.MatchedBy(emptySlice[flow.ServiceEvent]())).Return(exception).Once()
```

Comment:
This tripped me up for a while. In the past Testify didn't handle multiple `On` calls with the same method name and arguments (subsequent `On` calls used to be ignored), now it seems like they will handle it gracefully. Nice!